modules/rank-eval/src/test/java/org/elasticsearch/index/rankeval/RankEvalSpecTests.java

```diff
@@ -142,6 +146,104 @@ public void testRoundtripping() throws IOException {
+            types.add(randomAsciiOfLengthBetween(0, 50));
+        }
+
+        RankEvalSpec original = createTestItem();
+
+        List<NamedWriteableRegistry.Entry> namedWriteables = new ArrayList<>();
+        namedWriteables.add(new NamedWriteableRegistry.Entry(QueryBuilder.class, MatchAllQueryBuilder.NAME, MatchAllQueryBuilder::new));
+        namedWriteables.add(new NamedWriteableRegistry.Entry(RankedListQualityMetric.class, Precision.NAME, Precision::new));
+        namedWriteables.add(new NamedWriteableRegistry.Entry(
+                RankedListQualityMetric.class, DiscountedCumulativeGain.NAME, DiscountedCumulativeGain::new));
+        namedWriteables.add(new NamedWriteableRegistry.Entry(RankedListQualityMetric.class, ReciprocalRank.NAME, ReciprocalRank::new));
+
+
+        RankEvalSpec deserialized = RankEvalTestHelper.copy(original, RankEvalSpec::new, new NamedWriteableRegistry(namedWriteables));
+        assertEquals(deserialized, original);
+        assertEquals(deserialized.hashCode(), original.hashCode());
+        assertNotSame(deserialized, original);
+    }
+
+    public void testEqualsAndHash() throws IOException {
+        List<String> indices = new ArrayList<>();
+        int size = randomIntBetween(0, 20);
+        for (int i = 0; i < size; i++) {
+            indices.add(randomAsciiOfLengthBetween(0, 50));
+        }
+
+        List<String> types = new ArrayList<>();
+        size = randomIntBetween(0, 20);
+        for (int i = 0; i < size; i++) {
+            types.add(randomAsciiOfLengthBetween(0, 50));
+        }
+
+        RankEvalSpec testItem = createTestItem();
+
+        List<NamedWriteableRegistry.Entry> namedWriteables = new ArrayList<>();
+        namedWriteables.add(new NamedWriteableRegistry.Entry(QueryBuilder.class, MatchAllQueryBuilder.NAME, MatchAllQueryBuilder::new));
+        namedWriteables.add(new NamedWriteableRegistry.Entry(RankedListQualityMetric.class, Precision.NAME, Precision::new));
+        namedWriteables.add(new NamedWriteableRegistry.Entry(
+                RankedListQualityMetric.class, DiscountedCumulativeGain.NAME, DiscountedCumulativeGain::new));
+        namedWriteables.add(new NamedWriteableRegistry.Entry(RankedListQualityMetric.class, ReciprocalRank.NAME, ReciprocalRank::new));
+
+        RankEvalTestHelper.testHashCodeAndEquals(testItem, mutateTestItem(testItem),
+                RankEvalTestHelper.copy(testItem, RankEvalSpec::new, new NamedWriteableRegistry(namedWriteables)));
+    }
+
+    private RankEvalSpec mutateTestItem(RankEvalSpec original) {
+        Collection<RatedRequest> ratedRequests = original.getRatedRequests();
+        RankedListQualityMetric metric = original.getMetric();
+        Script template = original.getTemplate();
+
+        int mutate = randomIntBetween(0, 2);
+        switch (mutate) {
+            case 0:
+                Collection<RatedRequest> mutantRatedRequests = new ArrayList<>();
+                mutantRatedRequests.addAll(ratedRequests);
```

Comment:
If you had a copy like suggested above, there would be no need to copy over all old requests here.